File: src/osmpq/layout/cells.py
"""Quadtree cells over plain lon/lat, per docs/m0-contracts.md section 2.

The root cell covers lon [-180, 180], lat [-90, 90]. Each split produces
four children in Bing quadkey order: ``0`` = NW, ``1`` = NE, ``2`` = SW,
``3`` = SE. A cell key is the string of digits from the root; the root
itself is the key ``"root"``. Depth = number of digits (root = 0). Max
depth 20.

A point belongs to the cell where ``lon < east`` and ``lat < north`` are
strict on the high side (half-open intervals), except at the world's
eastern/northern edges, which are included in the last cell.

Internally every point is mapped to a 40-bit "quadkey code": the depth-20
digit string read as a base-4 integer. Because each digit is exactly 2 bits
and digits are ordered most-significant-first, the code of any cell at
depth ``d`` occupies a contiguous range of the 40-bit space, which is what
makes vectorized point-to-leaf assignment (and the builder's leaf-cell
selection over aggregated counts) possible without a per-row Python loop.
"""
from __future__ import annotations

from typing import Iterable, Sequence, Union

import numpy as np

MAX_DEPTH = 20
ROOT = "root"

# Bounding box in degrees, always (south, west, north, east).
BBox = tuple[float, float, float, float]

_WORLD = (-90.0, -180.0, 90.0, 180.0)


def _depth(key: str) -> int:
    return 0 if key == ROOT else len(key)


def cell_bbox(key: str) -> BBox:
    """Return (south, west, north, east) in degrees for a cell key."""
    if key == ROOT:
        return _WORLD
    south, west, north, east = _WORLD
    for digit in key:
        mid_lon = (west + east) / 2.0
        mid_lat = (south + north) / 2.0
        if digit == "0":  # NW
            west, east, south, north = west, mid_lon, mid_lat, north
        elif digit == "1":  # NE
            west, east, south, north = mid_lon, east, mid_lat, north
        elif digit == "2":  # SW
            west, east, south, north = west, mid_lon, south, mid_lat
        elif digit == "3":  # SE
            west, east, south, north = mid_lon, east, south, mid_lat
        else:
            raise ValueError(f"invalid quadkey digit {digit!r} in {key!r}")
    return (south, west, north, east)


def children(key: str) -> list[str]:
    """The four child keys of ``key``, in digit order 0,1,2,3."""
    if _depth(key) >= MAX_DEPTH:
        raise ValueError(f"cell {key!r} is already at max depth {MAX_DEPTH}")
    prefix = "" if key == ROOT else key
    return [prefix + d for d in "0123"]


def parent(key: str) -> str | None:
    """The parent key of ``key``, or None if ``key`` is the root."""
    if key == ROOT:
        return None
    if len(key) == 1:
        return ROOT
    return key[:-1]


def ancestors(key: str) -> list[str]:
    """Ancestors of ``key`` from its parent up to and including root.

    Does not include ``key`` itself. ``ancestors("root") == []``.
    """
    out: list[str] = []
    cur = parent(key)
    while cur is not None:
        out.append(cur)
        cur = parent(cur)
    return out


# --------------------------------------------------------------------------
# Quadkey codes (depth-20 integer encoding), scalar and vectorized
# --------------------------------------------------------------------------


def _grid_xy(lat: float, lon: float) -> tuple[int, int]:
    n = 1 << MAX_DEPTH
    x = int((lon + 180.0) / 360.0 * n)
    y = int((lat + 90.0) / 180.0 * n)
    x = min(max(x, 0), n - 1)
    y = min(max(y, 0), n - 1)
    return x, y


def point_to_qk(lat: float, lon: float) -> int:
    """Depth-20 quadkey code (40-bit int) for a point, MSB-first digits."""
    x, y = _grid_xy(lat, lon)
    inv_y = ((1 << MAX_DEPTH) - 1) ^ y
    code = 0
    for i in range(MAX_DEPTH - 1, -1, -1):
        bx = (x >> i) & 1
        biy = (inv_y >> i) & 1
        digit = (biy << 1) | bx
        code = (code << 2) | digit
    return code


def point_to_qk_np(lat_e7: np.ndarray, lon_e7: np.ndarray) -> np.ndarray:
    """Vectorized :func:`point_to_qk` over INT32 e7 coordinate arrays.

    Returns a uint64 array of 40-bit quadkey codes.
    """
    lat_e7 = np.asarray(lat_e7, dtype=np.int64)
    lon_e7 = np.asarray(lon_e7, dtype=np.int64)
    n = 1 << MAX_DEPTH
    lat = lat_e7.astype(np.float64) / 1e7
    lon = lon_e7.astype(np.float64) / 1e7
    x = np.floor((lon + 180.0) / 360.0 * n).astype(np.int64)
    y = np.floor((lat + 90.0) / 180.0 * n).astype(np.int64)
    np.clip(x, 0, n - 1, out=x)
    np.clip(y, 0, n - 1, out=y)
    inv_y = ((1 << MAX_DEPTH) - 1) ^ y
    code = np.zeros_like(x, dtype=np.uint64)
    for i in range(MAX_DEPTH - 1, -1, -1):
        bx = (x >> i) & 1
        biy = (inv_y >> i) & 1
        digit = (biy << 1) | bx
        code = (code << np.uint64(2)) | digit.astype(np.uint64)
    return code


def key_to_value(key: str) -> int:
    """The base-4 digit value of a key (0 for root)."""
    value = 0
    if key != ROOT:
        for ch in key:
            value = value * 4 + int(ch)
    return value


def qk_range(key: str) -> tuple[int, int]:
    """Inclusive [lo, hi] range of depth-20 quadkey codes covered by ``key``."""
    depth = _depth(key)
    value = key_to_value(key)
    shift = 2 * (MAX_DEPTH - depth)
    lo = value << shift
    span = 1 << shift
    return lo, lo + span - 1


class LeafIndex:
    """Precomputed structure over a leaf-cell set for fast lookups.

    Because the depth-20 quadkey code of every cell occupies a contiguous
    range of the 40-bit space, and the leaves partition the whole world,
    sorting leaves by their range's lower bound lets us find "the leaf
    containing this point" with a single vectorized ``searchsorted`` instead
    of a per-row descent.
    """

    def __init__(self, leaves: Iterable[str]):
        self.leaves: list[str] = sorted(set(leaves))
        self._leafset: frozenset[str] = frozenset(self.leaves)
        ranges = [(qk_range(k)[0], qk_range(k)[1], k) for k in self.leaves]
        ranges.sort(key=lambda t: t[0])
        self._los = np.array([r[0] for r in ranges], dtype=np.uint64)
        self._his = np.array([r[1] for r in ranges], dtype=np.uint64)
        self._keys_by_lo = [r[2] for r in ranges]

    def __contains__(self, key: object) -> bool:
        return key in self._leafset

    def __iter__(self):
        return iter(self.leaves)

    def __len__(self) -> int:
        return len(self.leaves)

    def leaf_index_for_qk(self, qk: np.ndarray) -> np.ndarray:
        """Index into ``self._keys_by_lo`` (== ``self.leaves_by_lo``) per code."""
        qk = np.asarray(qk, dtype=np.uint64)
        idx = np.searchsorted(self._los, qk, side="right") - 1
        np.clip(idx, 0, len(self._los) - 1, out=idx)
        return idx

    @property
    def leaves_by_lo(self) -> list[str]:
        return self._keys_by_lo

    def leaf_for_qk(self, qk: np.ndarray) -> np.ndarray:
        """Vectorized leaf key lookup; returns an object array of str."""
        idx = self.leaf_index_for_qk(qk)
        keys = np.array(self._keys_by_lo, dtype=object)
        return keys[idx]


Leaves = Union[set, frozenset, LeafIndex]


def _leaf_lookup(leaves: Leaves, key: str) -> bool:
    return key in leaves


def point_cell(lat: float, lon: float, leaves: Leaves) -> str:
    """The leaf cell (from ``leaves``) containing point (lat, lon)."""
    south, west, north, east = _WORLD
    key = ROOT
    depth = 0
    while True:
        if _leaf_lookup(leaves, key):
            return key
        if depth >= MAX_DEPTH:
            return key
        mid_lon = (west + east) / 2.0
        mid_lat = (south + north) / 2.0
        north_half = lat >= mid_lat
        east_half = lon >= mid_lon
        if north_half and not east_half:
            digit, west, east, south, north = "0", west, mid_lon, mid_lat, north
        elif north_half and east_half:
            digit, west, east, south, north = "1", mid_lon, east, mid_lat, north
        elif not north_half and not east_half:
            digit, west, east, south, north = "2", west, mid_lon, south, mid_lat
        else:
            digit, west, east, south, north = "3", mid_lon, east, south, mid_lat
        key = digit if key == ROOT else key + digit
        depth += 1


def point_cells_np(lat_e7: np.ndarray, lon_e7: np.ndarray, leaves: Leaves) -> np.ndarray:
    """Vectorized leaf-cell assignment for arrays of INT32 e7 coordinates.

    ``leaves`` may be a plain set/frozenset of leaf keys (converted to a
    :class:`LeafIndex` internally) or an existing :class:`LeafIndex`.
    Returns an object array of str leaf keys, one per point.
    """
    index = leaves if isinstance(leaves, LeafIndex) else LeafIndex(leaves)
    qk = point_to_qk_np(lat_e7, lon_e7)
    return index.leaf_for_qk(qk)


def _bbox_contains(outer: BBox, inner: BBox) -> bool:
    o_south, o_west, o_north, o_east = outer
    i_south, i_west, i_north, i_east = inner
    return i_south >= o_south and i_west >= o_west and i_north <= o_north and i_east <= o_east


def containing_cell(bbox: BBox, leaves: Leaves) -> str:
    """Smallest cell (leaf or ancestor, incl. root) fully containing ``bbox``.

    Descends from the root while exactly one child fully contains the whole
    bbox and that child is not itself a leaf (a leaf is where a branch of
    the actual tree stops, even if the bbox would geometrically fit deeper).
    """
    key = ROOT
    cur_bbox = _WORLD
    depth = 0
    while True:
        if _leaf_lookup(leaves, key):
            return key
        if depth >= MAX_DEPTH:
            return key
        south, west, north, east = cur_bbox
        mid_lon = (west + east) / 2.0
        mid_lat = (south + north) / 2.0
        candidates = {
            "0": (mid_lat, west, north, mid_lon),
            "1": (mid_lat, mid_lon, north, east),
            "2": (south, west, mid_lat, mid_lon),
            "3": (south, mid_lon, mid_lat, east),
        }
        contained = [d for d, cb in candidates.items() if _bbox_contains(cb, bbox)]
        if len(contained) != 1:
            return key
        digit = contained[0]
        key = digit if key == ROOT else key + digit
        cur_bbox = candidates[digit]
        depth += 1


def _bbox_intersects(a: BBox, b: BBox) -> bool:
    a_south, a_west, a_north, a_east = a
    b_south, b_west, b_north, b_east = b
    if a_east < b_west or a_west > b_east:
        return False
    if a_north < b_south or a_south > b_north:
        return False
    return True


DEFAULT_ANCESTOR_DEPTHS = [0, 3, 6, 9, 12]
DEFAULT_MAX_DEPTH_V2 = 13


def _value_to_key(value: int, depth: int) -> str:
    """Inverse of :func:`key_to_value` at a known ``depth`` (zero-padded)."""
    if depth == 0:
        return ROOT
    digits = []
    v = value
    for _ in range(depth):
        digits.append(str(v & 3))
        v >>= 2
    return "".join(reversed(digits))


def containing_cells_v2_np(
    ymin_e7: np.ndarray,
    xmin_e7: np.ndarray,
    ymax_e7: np.ndarray,
    xmax_e7: np.ndarray,
    leaves: Leaves,
    ancestor_depths: Sequence[int] = DEFAULT_ANCESTOR_DEPTHS,
    max_depth: int = DEFAULT_MAX_DEPTH_V2,
) -> np.ndarray:
    """Vectorized v2 loose-placement rule (docs/m1-contracts.md section 2).

    For each row's bbox (ymin/xmin/ymax/xmax in INT32 e7 units), find C, the
    smallest cell (leaf or ancestor, incl. root) that a plain unrestricted
    quadtree descent would stop at: this is exactly the cell identified by
    the longest common (quadkey-digit) prefix of the bbox's two extreme
    corners -- (ymin, xmin) and (ymax, xmax) -- truncated at ``max_depth``,
    and further truncated wherever the actual leaf set stops splitting that
    branch first. If that truncation stopped at an actual leaf, C is that
    leaf (used as-is, however deep). Otherwise C is an internal node of the
    real tree, so the result is the ancestor of C at the greatest depth in
    ``ancestor_depths`` that is <= depth(C).

    All inputs must be non-null (callers route null-bbox rows, e.g. ways
    with no resolvable geometry, to ``ROOT`` themselves before calling this).
    Returns an object array of ``str`` cell keys.
    """
    ymin_e7 = np.asarray(ymin_e7, dtype=np.int64)
    xmin_e7 = np.asarray(xmin_e7, dtype=np.int64)
    ymax_e7 = np.asarray(ymax_e7, dtype=np.int64)
    xmax_e7 = np.asarray(xmax_e7, dtype=np.int64)
    n = len(ymin_e7)
    if n == 0:
        return np.array([], dtype=object)
    if max_depth > MAX_DEPTH:
        raise ValueError(f"max_depth {max_depth} exceeds the module MAX_DEPTH {MAX_DEPTH}")

    index = leaves if isinstance(leaves, LeafIndex) else LeafIndex(leaves)

    # SW corner = (ymin, xmin), NE corner = (ymax, xmax): the bbox's two
    # extreme points. A quadtree cell is an axis-aligned rectangle, so it
    # fully contains the bbox iff it contains both of these corners.
    qk_sw = point_to_qk_np(ymin_e7, xmin_e7)
    qk_ne = point_to_qk_np(ymax_e7, xmax_e7)

    # Per-depth digit (0=MSB, i.e. depth-1 digit) of each corner's full
    # depth-MAX_DEPTH code, for depths 0..max_depth-1.
    digits_sw = np.empty((max_depth, n), dtype=np.uint8)
    digits_ne = np.empty((max_depth, n), dtype=np.uint8)
    for i in range(max_depth):
        shift = np.uint64(2 * (MAX_DEPTH - 1 - i))
        digits_sw[i] = ((qk_sw >> shift) & np.uint64(3)).astype(np.uint8)
        digits_ne[i] = ((qk_ne >> shift) & np.uint64(3)).astype(np.uint8)

    # Longest common prefix depth of the two corners' digit strings, capped
    # at max_depth (this is the depth of the smallest cell in an
    # unrestricted quadtree that contains both corners, hence the bbox).
    lcp_depth = np.full(n, max_depth, dtype=np.int64)
    still = np.ones(n, dtype=bool)
    for i in range(max_depth):
        mismatch = still & (digits_sw[i] != digits_ne[i])
        lcp_depth[mismatch] = i
        still &= ~mismatch

    # Depth of the actual leaf covering the SW corner (shared by the NE
    # corner too, for any depth <= lcp_depth, since they agree that far).
    leaf_idx = index.leaf_index_for_qk(qk_sw)
    leaf_depth_by_lo = np.array(
        [0 if k == ROOT else len(k) for k in index.leaves_by_lo], dtype=np.int64
    )
    leaf_depth = leaf_depth_by_lo[leaf_idx]

    truncated_depth = np.minimum(lcp_depth, leaf_depth)
    is_leaf_c = truncated_depth == leaf_depth

    ad_sorted = np.array(sorted(set(ancestor_depths)), dtype=np.int64)
    idx = np.searchsorted(ad_sorted, truncated_depth, side="right") - 1
    idx = np.clip(idx, 0, len(ad_sorted) - 1)
    rounded_depth = ad_sorted[idx]

    final_depth = np.where(is_leaf_c, truncated_depth, rounded_depth)

    # Build the full max_depth-digit string per row from the SW corner's
    # digits (shared with NE up to lcp_depth >= final_depth), then slice.
    acc = np.full(n, "", dtype=f"U{max_depth}")
    for i in range(max_depth):
        acc = np.char.add(acc, digits_sw[i].astype("U1"))

    final_depth_list = final_depth.tolist()
    out = np.empty(n, dtype=object)
    for row in range(n):
        d = final_depth_list[row]
        out[row] = ROOT if d == 0 else acc[row][:d]
    return out


def containing_cell_v2(
    bbox: BBox,
    leaves: Leaves,
    ancestor_depths: Sequence[int] = DEFAULT_ANCESTOR_DEPTHS,
    max_depth: int = DEFAULT_MAX_DEPTH_V2,
) -> str:
    """Scalar wrapper around :func:`containing_cells_v2_np` for a single bbox."""
    south, west, north, east = bbox
    ymin = np.array([round(south * 1e7)], dtype=np.int64)
    xmin = np.array([round(west * 1e7)], dtype=np.int64)
    ymax = np.array([round(north * 1e7)], dtype=np.int64)
    xmax = np.array([round(east * 1e7)], dtype=np.int64)
    result = containing_cells_v2_np(ymin, xmin, ymax, xmax, leaves, ancestor_depths, max_depth)
    return result[0]


def cells_for_bbox(bbox: BBox, leaves: Leaves) -> list[str]:
    """Leaves intersecting ``bbox`` plus all their ancestors (incl. root).

    This is the planner rule: "every leaf cell that intersects the bbox,
    plus every ancestor of those leaves up to and including root".
    """
    leaf_keys: Sequence[str] = leaves.leaves if isinstance(leaves, LeafIndex) else list(leaves)
    out: set[str] = set()
    for leaf in leaf_keys:
        if _bbox_intersects(cell_bbox(leaf), bbox):
            out.add(leaf)
            out.update(ancestors(leaf))
    out.add(ROOT)
    return sorted(out)

File: rust/osmpq-raw/src/cells.rs
//! Quadtree cells + Hilbert keys, matching `src/osmpq/layout/cells.py` and
//! `src/osmpq/layout/hilbert.py` bit-for-bit (see docs/m0-contracts.md
//! section 2 and docs/m1-contracts.md section 2).
//!
//! Cell keys are digit strings ("0213", ...), root = "root". Splits produce
//! four children in Bing quadkey order: 0=NW, 1=NE, 2=SW, 3=SE. A point
//! belongs to the cell where `lon < east` and `lat < north` are strict on
//! the high side (half-open), which is exactly what the bit-interleaved
//! "quadkey code" construction below gives you for free.

pub const ROOT: &str = "root";
pub const HILBERT_ORDER: u32 = 20;

/// ancestor_depths from docs/m1-contracts.md section 2.
pub const ANCESTOR_DEPTHS: [u32; 5] = [0, 3, 6, 9, 12];

/// Grid (x, y) coordinates at a given order, matching
/// `osmpq.layout.cells._grid_xy` / `point_to_qk_np` exactly (same floor/clip
/// arithmetic in f64).
#[inline]
pub fn grid_xy(lat_e7: i64, lon_e7: i64, order: u32) -> (u64, u64) {
    let n = (1u64 << order) as f64;
    let lat = lat_e7 as f64 / 1e7;
    let lon = lon_e7 as f64 / 1e7;
    let mut x = ((lon + 180.0) / 360.0 * n).floor() as i64;
    let mut y = ((lat + 90.0) / 180.0 * n).floor() as i64;
    let max_i = (1i64 << order) - 1;
    if x < 0 {
        x = 0;
    } else if x > max_i {
        x = max_i;
    }
    if y < 0 {
        y = 0;
    } else if y > max_i {
        y = max_i;
    }
    (x as u64, y as u64)
}

/// Depth-`order` quadkey code (2*order bits, MSB-first digits), matching
/// `point_to_qk` / `point_to_qk_np`.
#[inline]
pub fn qk_code(lat_e7: i64, lon_e7: i64, order: u32) -> u64 {
    let (x, y) = grid_xy(lat_e7, lon_e7, order);
    let mask = if order == 64 { u64::MAX } else { (1u64 << order) - 1 };
    let inv_y = mask ^ y;
    let mut code: u64 = 0;
    for i in (0..order).rev() {
        let bx = (x >> i) & 1;
        let biy = (inv_y >> i) & 1;
        let digit = (biy << 1) | bx;
        code = (code << 2) | digit;
    }
    code
}

/// Render a depth-`order` code as a cell key string ("" digits => "root").
pub fn code_to_key(code: u64, depth: u32) -> String {
    if depth == 0 {
        return ROOT.to_string();
    }
    let mut s = String::with_capacity(depth as usize);
    for i in (0..depth).rev() {
        let digit = ((code >> (2 * i)) & 0b11) as u8;
        s.push((b'0' + digit) as char);
    }
    s
}

/// The four child codes (in digit order 0,1,2,3) of a code at `depth`,
/// expressed at `depth+1`.
#[inline]
pub fn child_codes(code_at_depth: u64) -> [u64; 4] {
    let base = code_at_depth << 2;
    [base, base | 1, base | 2, base | 3]
}

/// Inclusive [lo, hi] range, at `max_depth` resolution, covered by a cell
/// whose code at `depth` is `value` (depth <= max_depth). Matches
/// `qk_range` generalized to a configurable resolution instead of the fixed
/// order-20 grid (the M1 histogram builds cells directly at `--max-depth`).
#[inline]
pub fn code_range_at(value: u64, depth: u32, max_depth: u32) -> (u64, u64) {
    let shift = 2 * (max_depth - depth);
    let lo = value << shift;
    let span = 1u64 << shift;
    (lo, lo + span - 1)
}

/// Classic Wikipedia `xy2d`: Hilbert curve index of (x, y) at `order`,
/// matching `hilbert_xy2d` exactly.
#[inline]
pub fn hilbert_xy2d(order: u32, mut x: u64, mut y: u64) -> u64 {
    let n: u64 = 1u64 << order;
    let mut d: u64 = 0;
    let mut s: u64 = n / 2;
    while s > 0 {
        let rx: u64 = if (x & s) > 0 { 1 } else { 0 };
        let ry: u64 = if (y & s) > 0 { 1 } else { 0 };
        d += s * s * ((3 * rx) ^ ry);
        if ry == 0 {
            if rx == 1 {
                x = n - 1 - x;
                y = n - 1 - y;
            }
            std::mem::swap(&mut x, &mut y);
        }
        s /= 2;
    }
    d
}

/// Hilbert key for a point given as INT32 e7 coordinates, matching
/// `hilbert_key` (fixed order 20, independent of the leaf max-depth).
#[inline]
pub fn hilbert_key(lat_e7: i32, lon_e7: i32) -> u64 {
    let (x, y) = grid_xy(lat_e7 as i64, lon_e7 as i64, HILBERT_ORDER);
    hilbert_xy2d(HILBERT_ORDER, x, y)
}

/// (south, west, north, east) world bbox.
const WORLD: (f64, f64, f64, f64) = (-90.0, -180.0, 90.0, 180.0);

/// One bisection step of `cur_bbox` toward digit `d` (0=NW,1=NE,2=SW,3=SE),
/// matching `cells_mod.cell_bbox` / `cells_mod.containing_cell` exactly
/// (same float ops, same order, so results match Python bit-for-bit).
#[inline]
fn bisect(cur: (f64, f64, f64, f64), digit: u8) -> (f64, f64, f64, f64) {
    let (south, west, north, east) = cur;
    let mid_lon = (west + east) / 2.0;
    let mid_lat = (south + north) / 2.0;
    match digit {
        0 => (mid_lat, west, north, mid_lon),
        1 => (mid_lat, mid_lon, north, east),
        2 => (south, west, mid_lat, mid_lon),
        3 => (south, mid_lon, mid_lat, east),
        _ => unreachable!(),
    }
}

/// outer fully contains inner (both (south, west, north, east)).
#[inline]
fn bbox_contains(outer: (f64, f64, f64, f64), inner: (f64, f64, f64, f64)) -> bool {
    let (o_south, o_west, o_north, o_east) = outer;
    let (i_south, i_west, i_north, i_east) = inner;
    i_south >= o_south && i_west >= o_west && i_north <= o_north && i_east <= o_east
}

/// A leaf cell set, indexed for fast lookups: leaves partition the whole
/// world into ranges of the depth-`max_depth` code space (same trick as
/// `osmpq.layout.cells.LeafIndex`).
pub struct LeafIndex {
    pub max_depth: u32,
    /// Sorted by `lo`.
    los: Vec<u64>,
    keys_by_lo: Vec<String>,
    depths_by_lo: Vec<u32>,
    leafset: std::collections::HashSet<String>,
}

impl LeafIndex {
    pub fn new(leaves: &[String], max_depth: u32) -> Self {
        let mut ranges: Vec<(u64, u64, String, u32)> = leaves
            .iter()
            .map(|k| {
                let depth = if k == ROOT { 0 } else { k.len() as u32 };
                let value = key_to_value(k);
                let (lo, hi) = code_range_at(value, depth, max_depth);
                (lo, hi, k.clone(), depth)
            })
            .collect();
        ranges.sort_by_key(|r| r.0);
        let los = ranges.iter().map(|r| r.0).collect();
        let keys_by_lo = ranges.iter().map(|r| r.2.clone()).collect();
        let depths_by_lo = ranges.iter().map(|r| r.3).collect();
        let leafset = leaves.iter().cloned().collect();
        LeafIndex {
            max_depth,
            los,
            keys_by_lo,
            depths_by_lo,
            leafset,
        }
    }

    pub fn contains(&self, key: &str) -> bool {
        self.leafset.contains(key)
    }

    pub fn leaves(&self) -> impl Iterator<Item = &str> {
        self.keys_by_lo.iter().map(|s| s.as_str())
    }

    /// The leaf cell (as an index into `keys_by_lo`) containing a
    /// depth-`max_depth` code.
    #[inline]
    fn leaf_idx_for_code(&self, code: u64) -> usize {
        // greatest lo <= code
        match self.los.binary_search(&code) {
            Ok(i) => i,
            Err(0) => 0,
            Err(i) => i - 1,
        }
    }

    #[inline]
    pub fn leaf_for_point(&self, lat_e7: i32, lon_e7: i32) -> &str {
        let code = qk_code(lat_e7 as i64, lon_e7 as i64, self.max_depth);
        let idx = self.leaf_idx_for_code(code);
        &self.keys_by_lo[idx]
    }

    #[inline]
    pub fn leaf_idx_for_point(&self, lat_e7: i32, lon_e7: i32) -> usize {
        let code = qk_code(lat_e7 as i64, lon_e7 as i64, self.max_depth);
        self.leaf_idx_for_code(code)
    }

    pub fn key_at(&self, idx: usize) -> &str {
        &self.keys_by_lo[idx]
    }

    pub fn depth_at(&self, idx: usize) -> u32 {
        self.depths_by_lo[idx]
    }

    /// Smallest cell (leaf or ancestor, including root) fully containing
    /// `bbox = (south, west, north, east)`, restricted (M1 v2 rule) to leaf
    /// cells and depths in `ANCESTOR_DEPTHS`.
    ///
    /// Descends from the root while exactly one child fully contains the
    /// whole bbox and the current cell is not a leaf; call the result C at
    /// depth d. If C is a leaf, use C. Otherwise use the ancestor of C at
    /// the greatest depth in `ANCESTOR_DEPTHS` that is <= d.
    pub fn containing_cell(&self, bbox: (f64, f64, f64, f64)) -> String {
        let mut key = ROOT.to_string();
        let mut cur_bbox = WORLD;
        let mut depth: u32 = 0;
        loop {
            if self.contains(&key) {
                return key;
            }
            if depth >= self.max_depth {
                return snap_to_ancestor_depth(&key, depth);
            }
            let mut contained_digit: Option<u8> = None;
            let mut n_contained = 0u32;
            let mut contained_bbox = cur_bbox;
            for digit in 0u8..4 {
                let child_bbox = bisect(cur_bbox, digit);
                if bbox_contains(child_bbox, bbox) {
                    contained_digit = Some(digit);
                    contained_bbox = child_bbox;
                    n_contained += 1;
                }
            }
            if n_contained != 1 {
                // Not exactly one child fully contains the bbox: C = key at
                // this depth.
                return snap_to_ancestor_depth(&key, depth);
            }
            let digit = contained_digit.unwrap();
            let digit_char = (digit + b'0') as char;
            if key == ROOT {
                key = digit_char.to_string();
            } else {
                key.push(digit_char);
            }
            cur_bbox = contained_bbox;
            depth += 1;
        }
    }
}

fn snap_to_ancestor_depth(key_at_stop: &str, depth_at_stop: u32) -> String {
    // `key_at_stop` is C (the cell we stopped at, non-leaf, depth
    // `depth_at_stop`). Use the ancestor at the greatest depth in
    // ANCESTOR_DEPTHS that is <= depth_at_stop.
    let target_depth = ANCESTOR_DEPTHS
        .iter()
        .copied()
        .filter(|&d| d <= depth_at_stop)
        .max()
        .unwrap_or(0);
    if target_depth == 0 || key_at_stop == ROOT {
        return ROOT.to_string();
    }
    key_at_stop[..target_depth as usize].to_string()
}

/// Select leaf cells from a dense histogram of depth-`max_depth` codes,
/// exactly as `osmpq.build.builder._select_leaf_cells` does (recursive
/// split-while-over-threshold), generalized to a configurable max depth
/// (the M1 change; M0 always used depth 20).
pub fn select_leaves(counts: &[u32], max_nodes_per_cell: u64, max_depth: u32) -> Vec<String> {
    let mut leaves = Vec::new();
    // cum[i] = sum(counts[0..i]); counts.len() == 4^max_depth
    let mut cum = vec![0u64; counts.len() + 1];
    for (i, c) in counts.iter().enumerate() {
        cum[i + 1] = cum[i] + *c as u64;
    }
    let range_count = |lo: u64, hi: u64| -> u64 { cum[(hi + 1) as usize] - cum[lo as usize] };

    fn recurse(
        value: u64,
        depth: u32,
        max_depth: u32,
        max_nodes: u64,
        range_count: &dyn Fn(u64, u64) -> u64,
        leaves: &mut Vec<String>,
    ) {
        let (lo, hi) = code_range_at(value, depth, max_depth);
        let n = range_count(lo, hi);
        if n <= max_nodes || depth >= max_depth {
            if n > 0 || (depth == 0 && value == 0) {
                leaves.push(code_to_key(value, depth));
            }
            return;
        }
        for cv in child_codes(value) {
            recurse(cv, depth + 1, max_depth, max_nodes, range_count, leaves);
        }
    }

    recurse(0, 0, max_depth, max_nodes_per_cell, &range_count, &mut leaves);
    if leaves.is_empty() {
        leaves.push(ROOT.to_string());
    }
    leaves
}

/// The base-4 digit value of a key (0 for root).
pub fn key_to_value(key: &str) -> u64 {
    if key == ROOT {
        return 0;
    }
    let mut value: u64 = 0;
    for ch in key.chars() {
        value = value * 4 + (ch as u64 - '0' as u64);
    }
    value
}
